Add proposal for Ansible webhooks (#2048)
operator-framework/operator-sdk: doc/proposals/ansible-operator-webhooks.md

## Release Signoff Checklist

- \[ \] Enhancement is `implementable`
- \[ \] Design details are appropriately documented from clear requirements
- \[ \] Test plan is defined
- \[ \] Graduation criteria for dev preview, tech preview, GA
- \[ \] User-facing documentation is created in [operator-sdk/doc][operator-sdk-doc]

## Open Questions (optional)

- For non-trivial playbooks, will the performance be acceptable?
- How much webhook configuration should we support?
- Should we support side-effects?
    Side-effects will require adding reconciliation mechanisms to the webhook,
    and would likely impact the performance even more extremely. My instinct is to not support it as it adds
    a lot of complexity and seems dangerous, but then we're obviously restricting the power of the webhooks.
    If we don't allow side-effects, we could enforce that by making the proxy reject requests from the webhooks
    that do anything but `GET`.
- Need to determine exactly what information we should send to the webhooks, and how it will be structured.
    Passing this information to the Runner will require some minor refactors.
- Where should artifacts for the webhook runs end up?

> 

Reviewer: If it has nothing to be added here then we should remove the section. 
However, may the timeout subject could be something raised here. WDYT?

## Summary

> This proposal defines a way for users to implement webhooks using Ansible playbooks/roles.

Reviewer: ```suggestion
Currently, Ansible-based operators have no webhook support at all, a user would have to write their own webhook server and deploy it alongside the operator to get that functionality. So, this proposal is to allow its users to implement webhooks using Ansible playbooks/roles.
```

## Motivation

- [Conversion webhooks](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion)
    are necessary to support CRD versioning
- [Mutating webhooks](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook)
    are necessary to support defaulting, though the addition of the default field to the OpenAPI spec may make this one 
    less important in the long-term.
- [Validating webhooks](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) 
    are necessary for higher level validation of incoming resources, for example SAR checks. There is an open community 
    issue explicitly requesting this feature: https://github.com/operator-framework/operator-sdk/issues/1658 .

> Currently Ansible-based operators have no webhook support at all, a user would have to write their own webhook server 

Reviewer: ```suggestion

```

and deploy it alongside the operator to get that functionality.

## Goals

> - Allow the creation and use of conversion/validating/mutating webhooks implemented with Ansible

Reviewer: The motivations lines one top shows that should be here as for example

```suggestion
- Support CRD versioning by using [Conversion webhooks](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion)
```

### Non-Goals
- Handle xWebhookConfiguration objects or associated Services
- Handle cert management

## Proposal

Add a webhooks section to the `watches.yaml`, that allows the user to specify lists of validating/mutating/conversion 
webhooks that should be created. 

The user should only need to provide a playbook or a role, and the url path for the webhook. 

The rest of the webhook configuration, ie the creation of the Service that points to the webhook server, the Webhook 
Configuration resources that lay out the rules and point to the proper Service, and the management of the certificates 
for the webhook server, will be the responsiblity of the user. 

The webhook configuration will look roughly as follows, with a big asterisk on the conversion webhooks which need some

> further investigation.

Reviewer: May add the points in the Risks and Risks and Mitigations

> # Ansible Operator Webhooks

Reviewer: WDYT about using the [template](https://github.com/operator-framework/operator-sdk/blob/master/doc/proposals/TEMPLATE.md)?

Author: I hadn't seen that before, it seems pretty heavy though, is there a lighter alternative or do we need to fill out the whole doc?

Reviewer: We can remove the sections which no make sense for the proposal, but I think we should follow up it.

Reviewer: Solved

> Status: Draft
    are necessary to support defaulting, though the addition of the default field to the OpenAPI spec may make this one 
## Background
- [Conversion webhooks](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion)
    are necessary to support CRD versioning
- [Mutating webhooks](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook)
    are necessary to support defaulting, though the addition of the default field to the OpenAPI spec may make this one 
    less important in the long-term.
- [Validating webhooks](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) 
    are necessary for higher level validation of incoming resources, for example SAR checks. There is an open community 
    issue explicitly requesting this feature: https://github.com/operator-framework/operator-sdk/issues/1658 .

Currently Ansible-based operators have no webhook support at all, a user would have to write their own webhook server 
and deploy it alongside the operator to get that functionality.

## Goals

> - Allow the creation and use of conversion/validating/mutating webhooks implemented with Ansible

Reviewer: IMHO the goals should be some which will describe the stories. E.g A operator user ansible developer should be able to ... in order to make it clear. WDYT?

## Non-Goals
- Handle xWebhookConfiguration objects or associated Services
- Handle cert management

> ## Solution

Reviewer: I think we should have here the description in story.. then each goal will be a story with its acceptance criteria. 
Note that we could use the acceptance criteria to define the tests as well. WDYT?

Add a webhooks section to the `watches.yaml`, that allows the user to specify lists of validating/mutating/conversion 
webhooks that should be created. 

The user should only need to provide a playbook or a role, and the url path for the webhook. 

The rest of the webhook configuration, ie the creation of the Service that points to the webhook server, the Webhook 
Configuration resources that lay out the rules and point to the proper Service, and the management of the certificates 
for the webhook server, will be the responsiblity of the user. 

The webhook configuration will look roughly as follows, with a big asterisk on the conversion webhooks which need some
further investigation.

```yaml
---
- version: v1alpha1
  group: apps.example.com
  kind: MyApp
  role: /opt/ansible/roles/myapp

>   webhooks:

Reviewer 2: Do we want to add a "timeout" such that we can cancel these processes early if they take too long?

Author: Possibly, the webhook resource also allows you to set a timeout on the request so ideally we'd want to match that, though I'm not sure how we would (I'll check if timeout is sent in the request object)

      path: /convert
```

The justification for allowing multiple webhooks per type is that webhooks can select objects by a variety of
attributes, so you could have a different webhook which is selected based on `metadata.labels`.

Based on the webhook configuration, during initialization we will create a webhook server and add it to manager,
and then register each of the webhooks to the webhook server, at the given path. We will create handlers that
respond to the requests by running the associated Ansible roles or playbooks.

- For validating webhooks, playbook success => Allowed, and playbook failure => Denied. We will report the
    failureMessages in the denial response.
- For mutating webhooks, we will create a custom module that will be used to report back the final desired
    object. We may need to write additional logic here (either in the module or on the Go-side) that generates
    a proper patch object from this.
- Conversion webhooks should work roughly the same as mutating webhooks.

## Discussion / Further Investigation
- For non-trivial playbooks, will the performance be acceptable?
- How much webhook configuration should we support?

> - Should we support side-effects?

Reviewer 2: My initial guess would be no. 

There are other ways to do the side effects (in the controller), and therefore we don't restrict the use cases IMO

    Side-effects will require adding reconciliation mechanisms to the webhook,
    and would likely impact the performance even more extremely. My instinct is to not support it as it adds
    a lot of complexity and seems dangerous, but then we're obviously restricting the power of the webhooks.
    If we don't allow side-effects, we could enforce that by making the proxy reject requests from the webhooks
    that do anything but `GET`.

> - Need to determine exactly what information we should send to the webhooks, and how it will be structured.

Reviewer 2: It might be nice to see what is sent to ansible runner for these in the proposal?

Author: I should clarify here, I mean how much of the request object, or should we just pass the whole request object in.
